### PR TITLE
[numpy] FFI random.exponential, logistic, gumbel, rayleigh, weibull, power, pareto

### PR DESCRIPTION
**Pull request by Yiyan66**
## Description ##
 FFI random.exponential, random.logistic, random.gumbel, random.rayleigh

 &#160;  | Old FFI (ctypes) (us) | New FFI (cython) (us)
:-------: | :-------------------: | :-: 
exponential       | 39.71               | 12.32
logistic       | 42.69                 | 12.6
gumbel       | 42.52                 | 12.56
rayleigh       | 39.67                | 12.13
weibull       | 39.53                | 12.17
power       | 32.09                 | 12.04
pareto       | 32.09                | 10.04

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/api/operator/numpy/random/np_exponential_op.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+MXNET_REGISTER_API("_npi.exponential")
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_exponential");
+  op::NumpyExponentialParam param;
+  nnvm::NodeAttrs attrs;
+  attrs.op = op;
+  if (args[1].type_code() == kDLInt) {
+      param.size = Tuple<int>(1, args[1].operator int64_t());
+  } else if (args[1].type_code() == kNull) {
```
Comment (hzfan) [inline src/api/operator/numpy/random/np_exponential_op.cc]:
```
} else if (args[1].type_code() == kNull) {
  param.size = dmlc::nullopt;
}
```

File: src/api/operator/numpy/random/np_exponential_op.cc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    param.scale = args[0].operator double();
+    attrs.parsed = std::move(param);
+    SetAttrDict<op::NumpyExponentialParam>(&attrs);
+    auto ndoutputs = Invoke(op, &attrs, 0, nullptr, &num_outputs, outputs);
+    if (out) {
+      *ret = PythonArg(3);
+    } else {
+      *ret = ndoutputs[0];
+    }
+  } else {
```
Comment (hzfan) [inline src/api/operator/numpy/random/np_exponential_op.cc]: Set `param.scale = dmlc::kNullOpt`

File: src/api/operator/numpy/random/np_exponential_op.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      param.size = Tuple<int>(1, args[1].operator int64_t());
+  } else if (args[1].type_code() == kNull) {
+      param.size = dmlc::nullopt;
+  } else {
+      param.size = Tuple<int>(args[1].operator ObjectRef());
+  }
+  if (args[2].type_code() != kNull) {
+    attrs.dict["ctx"] = args[2].operator std::string();
+  }
+  NDArray* out = args[3].operator mxnet::NDArray*();
```
Comment (hzfan) [inline src/api/operator/numpy/random/np_exponential_op.cc]: Seems we can make it more concise by sharing some code

```
NDArray* out = args[3].operator mxnet::NDArray*();
NDArray** outputs = out == nullptr ? nullptr : &out;
int num_outputs = out != nullptr;
NDArray* inputs[1];
int num_inputs = 0;
if (args[0].type_code() == kDLFloat || args[0].type_code() == kDLInt) {
  param.scale = args[0].operator double();
  num_inputs = 0;
} else {
  param.scale = dmlc::nullopt;
  inputs[0] = args[0].operator mxnet::NDArray*()
  num_inputs = 1;
}
attrs.parsed = std::move(param);
SetAttrDict<op::NumpyExponentialParam>(&attrs);
auto ndoutputs = Invoke(op, &attrs, num_inputs, inputs, &num_outputs, outputs);
if (out) {
  *ret = PythonArg(3);
} else {
  *ret = ndoutputs[0];
}

File: src/operator/numpy/random/np_power_op.h [hunk 4]
```diff
@@ -44,6 +45,7 @@ namespace op {
 struct NumpyPowerParam : public dmlc::Parameter<NumpyPowerParam> {
   dmlc::optional<float> a;
   dmlc::optional<mxnet::Tuple<int>> size;
+  std::string ctx;
```
Comment (hzfan) [inline src/operator/numpy/random/np_power_op.h]: Actually we do not need `ctx` here.

File: src/operator/numpy/random/np_power_op.h [hunk 5]
```diff
@@ -52,6 +54,17 @@ struct NumpyPowerParam : public dmlc::Parameter<NumpyPowerParam> {
       .describe("Output shape. If the given shape is, "
           "e.g., (m, n, k), then m * n * k samples are drawn. "
           "Default is None, in which case a single value is returned.");
+      DMLC_DECLARE_FIELD(ctx).set_default("cpu").describe(
```
Comment (hzfan) [inline src/operator/numpy/random/np_power_op.h]: Actually we do not need ctx here.

File: src/api/operator/numpy/random/np_exponential_op.cc (hunk 3 above)
Comment (hzfan) [inline src/api/operator/numpy/random/np_exponential_op.cc]: We may make it more concise.
```
NDArray* out = args[3].operator mxnet::NDArray*();
NDArray** outputs = out == nullptr ? nullptr : &out;
int num_outputs = out != nullptr;
NDArray* inputs[1];
int num_inputs = 0;
if (args[0].type_code() == kDLFloat || args[0].type_code() == kDLInt) {
  param.scale = args[0].operator double();
  num_inputs = 0;
} else {
  param.scale = dmlc::nullopt;
  inputs[0] = args[0].operator mxnet::NDArray*();
  num_inputs = 1;
}
attrs.parsed = std::move(param);
SetAttrDict<op::NumpyExponentialParam>(&attrs);
auto ndoutputs = Invoke(op, &attrs, num_inputs, inputs, &num_outputs, outputs);
if (out) {
  *ret = PythonArg(3);
} else {
  *ret = ndoutputs[0];
}

File: src/operator/numpy/random/np_power_op.h (hunk 4 above)
Comment (hzfan) [inline src/operator/numpy/random/np_power_op.h]: We do not need `ctx` here.

File: src/operator/numpy/random/np_power_op.h (hunk 5 above)
Comment (hzfan) [inline src/operator/numpy/random/np_power_op.h]: We do not need `ctx` here.

**Comment by Yiyan66** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by Yiyan66** [comment]
@mxnet-bot run ci [unix-gpu]

**Pull request closed without merging**